salt/grains/core.py

```diff
@@ -2098,6 +2202,10 @@ def get_machine_id():
     '''
     # Provides:
     #   machine-id
+
+    if platform.system() == 'AIX':
+        return {'machine_id': platform.machine()}
```

Comment:
Isn't `platform.machine()` returning the architecture of the machine? At least on Linux it does. I think `get_machine_id` should return something else here.

Reply:
I agree, platform.machine() usually returns the machine architecture, but apparently on AIX it does not, both with the Salt contained version of Python 2.7.5 that ships with Salt on AIX and with IBM's own Python 2.7.13.  On AIX the command uname -m displays the machine ID number of the hardware running the system, see https://www.ibm.com/support/knowledgecenter/ssw_aix_71/com.ibm.aix.cmds5/uname.htm

Hence the following sample:
root@l3ferp005_pub:~/test# python
Python 2.7.13 (default, Oct 17 2017, 05:37:56) [C] on aix6
Type "help", "copyright", "credits" or "license" for more information.
>>> import platform
>>> platform.machine()
'00F7F3FE4C00'
>>> platform.uname()  
('AIX', 'l3ferp005_pub', '1', '7', '00F7F3FE4C00', 'powerpc')
>>> 
root@l3ferp005_pub:~/test# uname -m
00F7F3FE4C00
root@l3ferp005_pub:~/test# 

Given Salt uses a self contained version of Python 2.7.5 which produces as above, believe it is safe to use for now, until AIX corrects their use of uname -m which normally returns the machine architecture on Linux.  It will have to be rechecked when Salt AIX supports Python 3 which gets more attention than Python 2.

Reply:
Isn't there perhaps a third way to get the machine ID that doesn't have the potential to suddenly break with an update?

Reply:
@dmurphy18 no, please. This is wrong. The `uname -m` does not give you a system ID, but the _hardware_ ID. Given AIX running on LPAR this simply won't work. Besides, it is deprecated. You should use `lsattr -El sys0` instead looking for system ID, which is typically appears as `os_uuid` if I remember it correctly after so many years.

Reply:
@dmurphy18 are we speaking of the same thing? The unique machine ID we are about, is the ID of **the local system** that is set during installation or boot, not a reference to the hardware (!!). That said, every time an admin reinstalls an OS on the _same_ hardware, he still supposed to get a different ID. As well as if one installs AIX on LPAR. And so this gives the whole different tracking of the entities in your asset ledger.

Now, if you want to use `uname -m`, you returning not a _system ID_, but _hardware ID_. This is why: its format is `xxyyyyyymmss` for a reason, where `xx` and `ss` are always zeroes and `mm` almost always set to `4C` as it is just a  ID of the CPU Planar. The rest `yyyyyy` is a Model ID of the hardware. Just like in your own example above. So it is kind of unique per a hardware (what we **do not** need), but per a software (what we **do** need). Now, if you will install AIX on LPAR, it will no longer return _the unique_ ID (while it _absolutely_ has to be as such), because LPAR will reference the same hardware, and thus you will render the whole feature **broken**. I am sure you already read the official manual from IBM and found the remark about this regarding LPAR, didn't you? Then you already got to the details why it is so and you already figured out that this is because it is the hardware id, not an OS id. And so you probably by now understood that you're returning the wrong type and thus got to the conclusion that `uname -m` is a bad idea after all.

Regarding `lsattr -El sys0` by "buggy", I mean that sometimes in rare cases you would need to use a different device for this to fetch the `os_uuid`. But sorry I do not remember AIX details, as it was really long years ago... But the `os_uuid` is exactly what you are looking for, if you care Salt's feature work as it supposed to.

Like I also said, alternatively, we can simply help almighty IBM to generate `/etc/machine-id` just like systemd does, as simply as using Python's internals:
```python
import uuid
print(uuid.uuid4().hex)
```

Reply:
@isbm Checked with Daniel internally since he originated this code and got clarification on the intent of this value for machine id (given documentation did not state equivocally what it represented), and now understand it as a value representing the machine/virtualization, similar to systemd/nspawn.  Using this interpretation the code needs to change, and possibly for Solaris too.  Will alter AIX code to do as advised.